### PR TITLE
Add basic wildcard support for listening to legacy activities.

### LINKED ISSUE #2183: Add wildcard support to AddLegacySource
Azure SDKs produce activities that have tags and links set and don't need extra enrichment. 

For client level spans we use the client and method name as activity name (i.e. `BlobClient.Download`) so it's impractical to use the existing `AddLegacySource(string name)` to collect them as customers would need to add every individual method they use.

Feature request:

Add wildcard support to AddLegacySource the same way it exists in AddSource so customers can write

```
.AddLegacySource("Azure.*");

### PR DESCRIPTION
**Pull request by BlakeWills**
Fixes #2183.

## Changes

Add support for wildcards for legacy (DiagnosticSource) events.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs [hunk 1]
```diff
@@ -172,13 +184,13 @@ internal sealed class TracerProviderSdk : TracerProvider
                     if (name.Contains('*'))
                     {
                         wildcardMode = true;
+                        break;
```
Comment (reyang) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:187]: 💯

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs [hunk 2]
```diff
@@ -71,7 +83,7 @@ internal sealed class TracerProviderSdk : TracerProvider
                     if (this.supportLegacyActivity && string.IsNullOrEmpty(activity.Source.Name))
```
Comment (reyang) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:83]: Given `supportLegacyActivity` is something we can determine ahead of time, consider extracting the logic and create the corresponding delegate with hard-wired logic, so we don't have to check it in every Activity start/stop.

```csharp

if (this.supportLegacyActivity)
{
    listener.ActivityStarted = (activity) =>
    {
        if (string.IsNullOrEmpty(activity.Source.Name))
        {
            // legacy logic
            // similar here, we already know whether wildcard mode should be used or not, we can hard-wire the logic
            return;
        }
        // new logic
    }
}
else
{
    listener.ActivityStarted = (activity) =>
    {
        // new logic
    }
}

Comment (reyang) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs:83]: Similar approach has been used here to optimize the hot path:
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/bd189aa654948e79203d66d4b6d33fb93557d7e9/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L185
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/bd189aa654948e79203d66d4b6d33fb93557d7e9/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L206

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs [hunk 3]
```diff
@@ -34,6 +34,8 @@ internal sealed class TracerProviderSdk : TracerProvider
         private readonly Sampler sampler;
         private readonly Action<Activity> getRequestedDataAction;
         private readonly bool supportLegacyActivity;
+        private readonly bool legacyActivityWildcardMode;
```
Comment (reyang) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:37]: I guess making this local and capture it in lambda/delegate would be faster than accessing the value via this pointer.
I haven't tried it myself, a [benchmark](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Trace/TraceBenchmarks.cs) or [stress test](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/test/OpenTelemetry.Tests.Stress) should give us the numbers.

**Review by reyang — APPROVED** [review]
LGTM, I put some perf suggestions.

**Comment by reyang** [comment]
@BlakeWills thanks for the contribution! Would you clear the EasyCLA so we can run the CI?

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 2 above)
Comment (BlakeWills) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: That makes sense and thanks for the tip. I've had a go at an initial optimization, let me know if this is what you expected.

Benchmark results (the top table is pre-optimisation, bottom is post), they're pretty marginal gains:

image

**Review by reyang — APPROVED** [review]
LGTM :shipit:

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 2 above)
Comment (reyang) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: 👍  :shipit:

**Review by cijothomas — APPROVED** [review]

**Pull request merged**